### PR TITLE
fix(cli): support mount path policy environment overrides

### LINKED ISSUE #859: Default coding-agent profile causes EXDEV for common tmp-to-final rename pattern
### 1. Minimal reproduce step (Required)

1. Mount a Drive9 path with the default FUSE profile (`coding-agent`):

   ```bash
   drive9 mount --mode=fuse :/workspace /mnt/drive9
   ```

2. Create a temporary file and a final destination directory on the mount:

   ```bash
   mkdir -p /mnt/drive9/repro/tmp
   mkdir -p /mnt/drive9/repro/blobs/aa
   printf 'payload\n' > /mnt/drive9/repro/tmp/object.blob
   ```

3. Invoke the rename syscall directly (do not use `mv`, which may fall back to copy-and-remove):

   ```bash
   python3 -c 'import os; os.rename("/mnt/drive9/repro/tmp/object.blob", "/mnt/drive9/repro/blobs/aa/object.blob")'
   ```

This was also reproduced with the exact CLI binary outside Kubernetes/CSI. The control run used the same CLI, endpoint, and paths with `--profile=none`.

### 2. What did you expect to see? (Required)

An atomic rename between two paths exposed by the same Drive9 mount should succeed, or the default profile should avoid placing a common temporary-to-final rename workflow across an implicit filesystem boundary.

### 3. What did you see instead (Required)

The default `coding-agent` profile returns `EXDEV`:

```text
OSError: [Errno 18] Invalid cross-device link
```

Observed reproduction matrix:

| Mount profile | Result |
| --- | --- |
| Default (`coding-agent`) | `errors.Is(err, syscall.EXDEV) == true`; source remains; target is absent |
| `none` | Rename succeeds and target content is correct |

The default profile includes `**/tmp/**` in `builtinCodingAgentLocalOnlyPatterns()`. Therefore the source is routed to the local-only layer while `blobs/**` remains remote. `Dat9FS.Rename` explicitly returns `EXDEV` when the source and target layers differ.

This confirms that the failure is caused by the built-in path-layer policy rather than a lack of rename support in the remote filesystem.

### 4. What is your drive9 version? (Required)

```text
component: drive9 mount
version: 99aceec
git_hash: 99aceec47c949c1b6f74233109cbdf8e10fb9d56
git_branch: HEAD
build_time: 2026-08-19T07:23:12Z
go_version: go1.25.1
```

## Root cause

The default mount profile creates an implicit backing-filesystem boundary based on the path name:

1. The default `coding-agent` profile includes `**/tmp/**` in `builtinCodingAgentLocalOnlyPatterns()`.
2. The rename source under `tmp/` is therefore classified as `PathLayerLocalOnly` and stored in the local overlay.
3. The destination under `blobs/` does not match a local-only rule and remains on the remote Drive9 layer.
4. `Dat9FS.Rename` resolves both path layers before performing the rename. If one path is local-only and the layers differ, it deliberately returns `syscall.EXDEV`.
5. The paths share one visible FUSE mount, but they do not share one Drive9 backing layer. The resulting `EXDEV` is therefore expected from the current layer policy, although the implicit `tmp` classification is surprising for atomic temp-to-final workflows.

Relevant code:

- `cmd/drive9/cli/profile.go`: the default `coding-agent` profile and `builtinCodingAgentLocalOnlyPatterns()`
- `pkg/fuse/local_policy.go`: the corresponding FUSE-side default policy
- `pkg/fuse/dat9fs.go`: `Dat9FS.Rename` returns `EXDEV` when source and destination layers differ

## Evidence supporting the root cause

1. The exact same CLI binary and endpoint reproduce `EXDEV` with the default profile.
2. Changing only the mount profile to `none` makes the same rename succeed with correct content.
3. The failure reproduces outside Kubernetes and CSI, excluding CSI subPath handling as the cause.
4. After `EXDEV`, the source remains and the target is absent, matching the early cross-layer guard in `Dat9FS.Rename`.

Therefore this is a Drive9 built-in path-policy issue, not a remote rename implementation failure, a real PVC device boundary, or a CSI-specific limitation.

### PR DESCRIPTION
**Pull request by mashenjun**
## Summary

- add newline-delimited `DRIVE9_MOUNT_LOCAL_ONLY_PATTERNS` and `DRIVE9_MOUNT_REMOTE_ONLY_PATTERNS` inputs
- merge environment patterns with profile rules and repeatable CLI flags
- snapshot environment-derived rules into supervised mount argv so restarts and adoption preserve the original policy
- clarify that `--remote-only` overrides local-only routing and add the `tmp` override example

## Problem

The default `coding-agent` profile classifies `**/tmp/**` as local-only. Renaming a temporary file from `tmp/**` into a remote-persistent path such as `blobs/**` therefore crosses backing layers and returns `EXDEV`.

This change keeps the built-in profile unchanged while giving callers an agent-legible way to force selected paths back to remote-persistent storage. Values use one pattern per line, are trimmed, deduplicated, validated with variable and line diagnostics, and follow the existing remote-over-local precedence.

Example:

```bash
export DRIVE9_MOUNT_REMOTE_ONLY_PATTERNS=$'**/tmp/**\n**/.tmp/**'
drive9 mount :/workspace /mnt/drive9
```

## Validation

- `go test ./cmd/drive9/cli ./cmd/drive9 -count=1`
- `go test ./pkg/fuse -run 'TestLocalPolicy(RemoteOnlyOverridesBuiltinTmpRule|RemoteOnlyOverridesLocalOnly)$' -count=1`
- `go vet ./cmd/drive9/cli ./pkg/fuse ./cmd/drive9`
- `git diff --check`

Local full-package FUSE validation was attempted, but the host had approximately 0.5% free disk space, below the suite's 10% write-back threshold, so unrelated quota tests returned `ENOSPC`. Local `make lint` was also blocked because the repository-pinned golangci-lint binary was built with Go 1.24.6 while this module targets Go 1.25.1.

## Follow-up

CSI propagation is intentionally deferred to [drive9-ai/k8s-csi#36](https://github.com/drive9-ai/k8s-csi/issues/36).

Closes #859

## Summary by CodeRabbit

* **New Features**
  * Added environment-variable support for configuring local-only and remote-only mount path patterns.
  * Patterns are validated, deduplicated, and applied to foreground, background, and supervised mounts.
  * Remote-only rules can override built-in local-only rules where applicable.

* **Documentation**
  * Updated mount help text with environment-variable details and usage examples.

* **Bug Fixes**
  * Prevented mount policy environment variables from leaking into child processes after being consumed.

### CODE REVIEW

File: cmd/drive9/cli/mount.go [hunk 1]
```diff
@@ -304,6 +308,25 @@ func fsMountCmdWithBackground(args []string, background bool) error {
...
+	policyEnvArgs := make([]string, 0, 2*(len(envLocalOnlyPatterns)+len(envRemoteOnlyPatterns)))
+	for _, pattern := range envLocalOnlyPatterns {
+		policyEnvArgs = append(policyEnvArgs, "--local-only", pattern)
+	}
+	for _, pattern := range envRemoteOnlyPatterns {
+		policyEnvArgs = append(policyEnvArgs, "--remote-only", pattern)
+	}
+	// Snapshot environment-derived policy in argv so supervised restarts and
+	// later adoption keep the original mount contract without re-reading env.
+	originalArgs = append(policyEnvArgs, originalArgs...)
```
[the remark this answers is not in the document]
Comment (mashenjun) [reply cmd/drive9/cli/mount.go:340]: Fixed in 9252fd25. The legacy background branch now passes the materialized originalArgs, preserving environment-derived path-policy flags for --no-supervise and DRIVE9_MOUNT_SUPERVISE=off. The existing legacy-background test now asserts both remote-only patterns reach the child argv.

**Comment by mashenjun** [comment]
@codex review

File: cmd/drive9/cli/mount.go [hunk 2]
```diff
@@ -304,6 +308,25 @@ func fsMountCmdWithBackground(args []string, background bool) error {
...
+	envRemoteOnlyPatterns, err := consumeMountPolicyPatternsEnv(EnvMountRemoteOnlyPatterns)
+	if err != nil {
+		return err
+	}
+	policyEnvArgs := make([]string, 0, 2*(len(envLocalOnlyPatterns)+len(envRemoteOnlyPatterns)))
+	for _, pattern := range envLocalOnlyPatterns {
+		policyEnvArgs = append(policyEnvArgs, "--local-only", pattern)
+	}
+	for _, pattern := range envRemoteOnlyPatterns {
+		policyEnvArgs = append(policyEnvArgs, "--remote-only", pattern)
```
[the remark this answers is not in the document]
Comment (mashenjun) [reply cmd/drive9/cli/mount.go]: Fixed in caf49fb8. Environment-derived rules now materialize as single --local-only=<pattern> and --remote-only=<pattern> argv tokens. Regression coverage verifies that --api-key survives legacy credential stripping and --foreground survives supervised worker argument filtering.

**Comment by mashenjun** [comment]
@codex review

**Review by srstack — APPROVED** [review]
LGTM. I reviewed the full diff, checked out the branch, and verified locally:

- `go test ./cmd/drive9/cli ./cmd/drive9 -count=1` — pass
- `go test ./pkg/fuse -run 'TestLocalPolicy(RemoteOnlyOverridesBuiltinTmpRule|RemoteOnlyOverridesLocalOnly)$' -count=1` — pass
- `go vet ./cmd/drive9/cli ./pkg/fuse ./cmd/drive9` — clean

The design is solid:

- **Argv snapshotting** is the right call: consuming the env vars once, materializing them as `--local-only=`/`--remote-only=` args on `originalArgs`, and scrubbing the env (both via `os.Unsetenv` in `consumeMountPolicyPatternsEnv` and via `mountBackgroundEnv`/`applyScrubbedMountEnv`) guarantees supervised restarts and adoption re-derive the identical policy without double-applying it.
- **Flag-injection safety** is handled correctly: patterns are passed in `--flag=value` form, so values like `--api-key` or `--foreground` cannot be reinterpreted as flags by the re-exec'd child. Good to see explicit regression tests for this (`TestMountCmdNoSuperviseUsesLegacyBackground`, `TestMountCmdMaterializesPolicyEnvironmentInSupervisedArgs`).
- **Precedence** is preserved: remote-only is checked before local-only in `LocalPolicy.classifyWithSource`, and env patterns merge between profile defaults and CLI flags via the existing variadic `mergeProfileValues`, so `DRIVE9_MOUNT_REMOTE_ONLY_PATTERNS='**/tmp/**'` cleanly overrides the built-in `coding-agent` tmp rule — the exact EXDEV rename scenario from #859.
- **Diagnostics** are operator-friendly: validation errors name the variable and line number.
- Vault and object-mount paths are unaffected: the env vars never reach their background children because `mountBackgroundEnv` strips them, and `foregroundMountCommandArgs`'s `args[0] == "vault"` special case can't be confused by the prepended policy flags since vault mounts never flow through `fsMountCmdWithBackground`.

Two non-blocking nits:

- [NIT] In the object-store branch (the early return before policy env consumption), the policy env vars remain set in the parent process env. Harmless since they're stripped from any spawned child, but a one-line comment noting the intent would help future readers.
- [NIT] `patterns := make([]string, 0)` in `consumeMountPolicyPatternsEnv` could just be `var patterns []string` — returning `nil` for the empty case is equally fine for all callers.

Nice work — approving.

**Review by qiffang — APPROVED** [review]
Reviewed exact head `caf49fb8ea68a0e2b7d19fd9eaf4d8accb6e04f0` end to end. I found no code-level blocker.

The policy contract closes correctly:

- both environment variables are consumed once, split by line, trimmed, deduplicated, and validated with variable/line diagnostics;
- effective ordering is profile rules, then environment rules, then explicit repeatable CLI flags, while FUSE still evaluates remote-only before local-only, so `**/tmp/**` really overrides the built-in coding-agent local rule;
- environment rules are materialized as single `--local-only=<pattern>` / `--remote-only=<pattern>` argv tokens before supervision, then scrubbed from child environments. This preserves the original policy across normal supervision, restart/adoption, `--supervise-foreground`, and legacy `--no-supervise` without double-applying ambient environment changes;
- single-token materialization keeps values such as `--api-key` and `--foreground` opaque to credential/control-flag filters;
- object mounts do not consume the overlay policy, and unsupported non-overlay/WebDAV combinations still fail rather than silently pretending to honor it.

Validation passed locally: `git diff --check`, touched-file gofmt, `go test ./cmd/drive9/cli ./cmd/drive9 -count=1`, focused FUSE policy tests, focused race tests, `go vet`, Windows CLI test compilation, and `make lint`. Exact-head hosted Code CI and Local E2E are green.

Merge-state caveat: GitHub currently reports `CONFLICTING` / `DIRTY` against main `d6bdc83`; main added the gVisor mount flag/env in the same `mount.go` and test regions. Rebase before merge, preserve both contracts (gVisor env behavior and policy-env argv snapshot/scrub), rerun required checks, and request a delta review on the new exact head.

**Review by mornyx — COMMENTED** [review]
## Review summary

This change adds newline-delimited `DRIVE9_MOUNT_LOCAL_ONLY_PATTERNS` / `DRIVE9_MOUNT_REMOTE_ONLY_PATTERNS`, merges them as profile → env → CLI, snapshots them as single `--local-only=<pattern>` / `--remote-only=<pattern>` argv tokens, and scrubs the variables from child environments so supervise, adopt, and legacy `--no-supervise` restarts cannot double-apply. Under a strict maintainability bar the overlay contract is correct on current HEAD: both prior Codex findings are gone, FUSE still classifies remote-only before local-only so `**/tmp/**` overrides the coding-agent tmp rule, and object/vault mounts do not silently honor overlay policy. One should-fix remains on the new ambient channel: non-overlay / WebDAV Drive9 mounts now hard-fail with a diagnostic that names CLI flags the caller never passed.

**Resolved since last review**
- Legacy `--no-supervise` / `DRIVE9_MOUNT_SUPERVISE=off` now spawns from materialized `originalArgs`, so env-derived rules reach the child argv (`cmd/drive9/cli/mount.go` around the `startMountBackground` call; covered by `TestMountCmdNoSuperviseUsesLegacyBackground`).
- Env patterns materialize as `--flag=value` tokens, so values such as `--api-key` and `--foreground` survive `stripBackgroundOnlyCredentialArgs` and `workerArgsForSupervise`.
- The earlier main-conflict caveat on the gVisor mount flag/env in the same region is no longer current; GitHub reports the PR mergeable, and the two contracts do not clobber each other.

**What looks good**
- Consume-once, argv snapshot, then env scrub is the right persistence model for `drive9 mount ensure` / adopt: the original routing rules live in sanitized args, not in a later shell's environment.
- Validation names the variable and 1-based line, and the adversarial tests actually exercise the flag-injection boundary instead of only happy-path patterns.
- The built-in `coding-agent` profile is left unchanged; the #859 EXDEV case is pinned at the FUSE layer (`TestLocalPolicyRemoteOnlyOverridesBuiltinTmpRule`).

---

### Should fix (non-blocking, strong)

**1. Non-overlay / WebDAV failures misattribute env policy to CLI flags**

Env-derived patterns are merged into `effectiveLocalOnlyPatterns` / `effectiveRemoteOnlyPatterns` for every Drive9 mount (`cmd/drive9/cli/mount.go` around the `mergeProfileValues(..., envLocalOnlyPatterns, ...)` call). `validateMountProfileFlags` then treats a non-empty list as `hasPolicyFlags` and, for `interactive` / `none` (including implicit WebDAV profile selection), returns:

`drive9 mount: --local-root, --local-only, --remote-only, and pack paths require an overlay profile`

Failing closed rather than pretending to honor overlay policy is the right call, and it matches the existing CLI-flag behavior. Two problems with the new env channel:

- The error never names `DRIVE9_MOUNT_LOCAL_ONLY_PATTERNS` / `DRIVE9_MOUNT_REMOTE_ONLY_PATTERNS`. The automation audience this feature targets (exported env, direnv, agent harness) will look for flags they did not pass.
- Object-store and vault mounts still ignore the same env (early return before consume; children scrubbed), so the same export is silent on one non-overlay surface and a hard failure on another.

Keep the strictness. Track provenance so env-only contributions produce a message that names the variable (for example, `DRIVE9_MOUNT_REMOTE_ONLY_PATTERNS is set but --profile=none cannot honor overlay policy`), and add a test that pins env-policy behavior for `--profile none` and `--mode=webdav`. If object-store/vault should stay silent, say so next to the early return.

---

### Nits

- `drive9 mount systemd-unit` is a flag pass-through generator (`runMountSystemdUnit` copies `rest` into `ExecStart` and never consumes these variables). That matches how other ambient mount env (API key, gVisor) is handled: bake `--remote-only=` into the unit, or set `Environment=` on the service. Generating a unit from a shell that only exported the new variables will silently drop the override. Not unique to this PR; worth a one-line help note if systemd is a supported persistence path for agent mounts.
- CLI two-token form (`--local-only --foreground`) still interacts with `workerArgsForSupervise` / `stripBackgroundOnlyCredentialArgs` the way the env path used to. Pre-existing; the env input is fixed. A later pass could normalize every `--local-only` / `--remote-only` pair to `=` form at snapshot time.
- `validateMountPolicyPatterns` still says `invalid local policy pattern` for remote-only env lines (`DRIVE9_MOUNT_REMOTE_ONLY_PATTERNS line 2: ... invalid local policy pattern`). Reword to "invalid mount policy pattern".
- `patterns := make([]string, 0)` in `consumeMountPolicyPatternsEnv` can be `var patterns []string`. Object-store early return leaves the parent process env set; children are already scrubbed.

---

### Recommendation
**Comment.** Overlay snapshot/scrub is correct and the two prior regressions are gone; the misattributed non-overlay env failure is should-fix, not a merge block and not an LGTM.
